libs/gl-client/src/node/mod.rs

```diff
@@ -23,6 +25,86 @@ pub trait GrpcClient {
     fn new_with_inner(inner: service::AuthService) -> Self;
 }
 
+/// A builder to set up and configure a `Node`.
+pub struct NodeBuilder {
+    node_id: Vec<u8>,
+    network: Network,
+    tls: Option<TlsConfig>,
+    rune: Option<futhark::Rune>,
+    auth: Option<Vec<u8>>,
+}
+
+impl NodeBuilder {
+    /// Sets a `TlsConfig` for the `Node`. Overrides a `TlsConfig` created from
+    /// the auth blob using `with_auth`.
+    pub fn with_tls(self, tls: TlsConfig) -> Self {
+        Self {
+            node_id: self.node_id,
+            network: self.network,
+            tls: Some(tls),
+            rune: self.rune,
+            auth: self.auth,
+        }
+    }
+
+    /// Sets a `futhark::Rune` for the `Node`. Overrides a `futhark::Rune` that
+    /// is extracted from the auth blob via `with_auth`.
+    pub fn with_rune(self, rune: futhark::Rune) -> Self {
+        Self {
+            node_id: self.node_id,
+            network: self.network,
+            tls: self.tls,
+            rune: Some(rune),
+            auth: self.auth,
+        }
+    }
+
+    /// Use the auth blob to create the `TlsConfig` and the `futhark::Rune` for
+    /// the `Node`. Will be overridden by `with_rune()` and `with_tls`.
+    pub fn with_auth(self, auth: &[u8]) -> Self {
+        Self {
+            node_id: self.node_id,
+            network: self.network,
+            tls: self.tls,
+            rune: self.rune,
+            auth: Some(auth.to_vec()),
+        }
+    }
+
+    /// Build the `Node` from set parameters.
+    pub fn build(self) -> Result<Node> {
+        let mut tls = None;
+        let mut rune = None;
+
+        if let Some(auth) = self.auth {
+            // deserialize auth blob and create identity;
+            let blob = serialize::AuthBlob::deserialize(&auth[..])?;
+            tls = Some(TlsConfig::new()?.identity(blob.cert, blob.key));
+            let srune = Rune::from_base64(&blob.rune)?;
+            rune = Some(srune);
+        }
+
+        if let Some(stls) = self.tls {
+            tls = Some(stls);
+        }
+
+        if let Some(srune) = self.rune {
+            rune = Some(srune);
+        }
+
+        if tls.is_none() || rune.is_none() {
+            return Err(anyhow!("both, tls and rune must be present, or auth data"));
```

Comment:
The thing is, nothing here is really optional, it is just a matter of convenience ^^. You can either create a node from `auth` alone, or from the tuple `TlsConfig` and a `Rune`. The goal here was to give some flexibility while keeping backwards compatibility (auth is a braking change otherwise).

Maybe the best idea would be to have several methods that return the `NodeBuilder`, one that is called `from_auth` and one that is called `from_tls` as a starting point with the option to override those by other `with_xyz` methods before finalizing?